### ISSUE TITLE
DEComplexDisease

### ISSUE THREAD

**Issue by menggf** [Bioconductor/Contributions#455]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/menggf/DEComplexDisease

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by nturaga** [comment]
# Review DEComplexDisease 

## R CMD build 

* ok

## R CMD INSTALL

* ok

## DESCRIPTION

* ok.

* DifferntialExpression as a biocView.

## NAMESPACE

* Add

    ```
    importFrom("grDevices", "rainbow")
    importFrom("graphics", "abline", "axis", "legend", "lines", "mtext",
             "par", "plot", "points", "text")
    importFrom("methods", "is")
    importFrom("stats", "fisher.test", "kmeans", "lm", "loess", "pnbinom",
             "pnorm", "predict", "quantile", "sd")
    ```

## man

* ok.

## vignette

* Good vignette as far as information.

* Just keep the Rmd file, get rid of everything else.

* Make sure your code chunks evaluate, use 

    ```
    {r, eval=TRUE}
    ```

* Use BiocFileCache, to `download.file("https://...")` in vignette. FYI, link is dead.


## C++ 

* Instead of sig.cpp, l:530

```
 cout<<"DEG: "<< st <<":"<< num_w << "\t" << x << "\t" << y << "\t" << overlap << "\t" <<         output1 <<endl;
```

use below or remove the above code.

```
Rcpp::Rcout << .....
```


## R


* util.R is totally unreadable code. There is no documentation, and the variable names are poorly constructed.

* Please use formatR to format your code(https://cran.rstudio.com/web/packages/formatR/index.html). Eg: Plot.deg, 
  the indentation is off, which makes your code very hard to read.

  Make this change in all your `.R` files.


*  instead of 

```
  88                 bi=sapply(1:length(p), function(w){
  89                         if(p[w] <= mycutoff[w] )
  90                             return(1)
  91                         if(1- p[w] <= mycutoff[w] )
  92                             return(-1)
  93                         return(0);
  94                     });

```

Use:

```
bi = integer(length(p))
bi[ p <= mycutoff ] = 1
bi[ 1 - p <= mycutoff ] = -1
```


* Use `BiocParallel::bplapply` instead of `mclapply` everywhere.

* Avoid

    ```
    module.screen.R
    72                 p=sapply(1:length(c.feature), function(i){
    ```

    Use `seq_along(c.feature)` instead and similar. Wherever you see,
    `1:length(blah)`, please use `seq_along` or `seq_len` as needed.

    Similarly in module.compare.R, l 69,

    ```
     for(i in 1:n1){
         ges1=res.module1[[mod1[i]]][[type]][["genes"]];
         for(j in 1:n2){
             ges2=res.module2[[mod2[j]]][[type]][["genes"]];
             olp2[i,j]=length(intersect(ges1, ges2));
             olp4[i,j]=olp2[i,j]/max(len3[i], len4[j])
         }
     }
     ```

     Use seq_len(n1) and seq_len(n2) as well. 
     Also, check, module.overlap.R, module.screen.R for the same issue.

    if you want to iterate from the 10th element onwards eg: in module.scree.R, l 56

    ```
    tail(seq_along(add.patients),-9)
    ```

* In summarize.module.R,

```
pas=pas[pas != "decd.input" & pas != "decd.speicific" & pas != "decd.clustering"];
```
You have a typo for `specific`.

* Closely examine ALL iterations (`for`, `lapply()`, `sapply()`, `apply()`, ...) for opportunities to "vectorize", i.e., replace many calls to R functions with one call, as in the following, from module.modeling.R:118

```
newsc=apply(sub.deg,2, function(x) length(which(x== new.seed))/new.n)
```

has `ncol(sub.deg)` calls to `length(which(.))`, `==`, and `/`, whereas

```
newsc = colSums(sub.deg == new.seed) / new.n
```

has just 1 call to `colSums()`, `==`, and `/`

Do this everywhere.

* Replacing iterations with vectorization will greatly speed up your code. Maybe with vectorization it will NOT be necessary to use parallel evalation?

* summarize.module.R: use a 'helper' function to simplify these sapply(), e.g.,

```
    fun = function(x, elt1, elt2) res.module[[x]][[elt1]][[elt2]]
    sc1 = unlist(sapply(pas, fun, "max.genes", "sc"))
    ...
```

  but these "nested lists" seem to be a problem throughout your code, e.g.,  module.overlap.R:50. A better approach is to make your data 'tidy' when you create it, instead of trying to manipulate complicated nested lists. See http://vita.had.co.nz/papers/tidy-data.html.

**Comment by menggf** [comment]
Dear @nturaga ,
Thank you for the good comments. Please give more time to do the modification. I hope to submit the new submission before the end of next week (Sep. 8).

Best Regards,
Guofeng,

**Comment by nturaga** [comment]
Hi @menggf 

Let me know when you are ready for the next review.

**Comment by menggf** [comment]
Hi @nturaga ,

I corrected nearly all the mentioned problems and some other problems.
Best, Guofeng,

**Comment by nturaga** [comment]
Hi @menggf 

 The vignette doesn't evaluate because the R tags are not used. I need to be able to evaluate the vignette before i'm able to review your code again.

 https://www.bioconductor.org/developers/package-guidelines/#vignettes

 Please follow the vignette guidelines, also check other bioconductor packages for vignettes if you are not sure how to write them.

**Comment by nturaga** [comment]
* NAMESPACE edited by hand, instead of adding @importFrom at the appropriate places in the               documentation.

     Please check http://r-pkgs.had.co.nz/namespace.html,

 "Roxygen2 makes NAMESPACE tidy. No matter how many times you use @importFrom foo bar you’ll only get     one importFrom(foo, bar) in your NAMESPACE. This makes it easy to attach import directives to every      function that need them, rather than trying to manage in one central place."

 * Vignette doesn't evaluate because the R tags are not used. The reviewers need to be able to evaluate   the vignette, the package cannot go through without this.

 https://www.bioconductor.org/developers/package-guidelines/#vignettes

 Please follow the vignette guidelines, also check other bioconductor packages.

 e.g

 \`\`\`{r, eval=TRUE}
  hc=hclust(dist(t(exp.brca[, cl==0])))
  \`\`\`

 * Remove commented out code

**Comment by nturaga** [comment]
@menggf Please respond with the changes you have made. I will close the issue if there is no response by Oct 19th (2 weeks from this message).

**Comment by menggf** [comment]
Hi, @nturaga ,

Sorry for the laggard response for the reason that I am on a holiday. 

Most of the changes are done by following your comments, including but not limited to:
(1) update the package dependence in Description.txt and Roxygen2 document
(2) Add comments to util.R
(3) Vectorize some "for" and "apply" statement, including all the mentioned comments and several new ones.
(4) Replace mcapply with BiocParallel::bplapply
(5) Use seq_along or seq_len to generate sequential vectors
(6) Format the code usingformatR

Not changed:
vignette  code chunks evaluate: the vignette need huge dataset to output the same results and it is not feasible.

Best Regards,
Guofeng,

**Comment by nturaga** [comment]
Hi @menggf 

* It's `Depends`, not `Dependence` for the DESCRIPTION file tag.  Please correct this.

* The vignette is important for a bioconductor package, it needs to be evaluated.  Can't you make the data set smaller for the vignette? There are helpful guidelines here http://bioconductor.org/developers/package-guidelines/#vignettes.  If a user cannot run your vignette in a meaningful way, then they will have trouble following your package or using it.

**Comment by menggf** [comment]
Hi, @nturaga ,

We completely understand your concerns and also believe it important to make vignette to be evaluated. Indeed, we have attempted to select the samples and genes to generate vignette data with acceptable size. However, the analysis results were not right and it would mislead readers. We believe that it is completely not acceptable to generate improper results while just to make sure it technologically sounds. 

Best Regards,

Guofeng,

**Comment by mtmorgan** [comment]
The purpose of the vignette is to illustrate package use, not to replicate a comprehensive analysis. NO OTHER BIOCONDUCTOR PACKAGE AUTHOR has failed to provide a meaningful vignette with most code chunks fully evaluated. It is a REQUIREMENT of Bioconductor packages to have such a vignette, so you must either identify a suitable approach or find other venues for distributing your package.

A typical problem is that code is inefficient, the nested 'for' loops in your code make it seem likely that you are using iteration rather than vectorization. Two useful resources on this are http://bioconductor.org/developers/how-to/efficient-code/#vectorize and https://github.com/Bioconductor/Contributions/issues/516#issuecomment-337962416 .

**Comment by lawremi** [comment]
Technical soundness needs to take precedence over whether the results are interesting/meaningful. One compromise is demonstrate an uninteresting but working analysis, and then load cached results from a bigger one to demonstrate downstream steps.

**Comment by menggf** [comment]
@nturaga , @mtmorgan , @lawremi ,
If I did not misunderstand it, R package is technical evaluated in two ways: (1) \examples{} in *.Rd and (2) vignette chunks. We use the former as technical evaluation for every function of this package. All the examples have been set runnable and they will generate meaningless results.  We use the later to demonstrate package usage and explain the users how to interpret the results. It needs users to download dataset from third-party websites.  So, we set it to be not evaluated. 

If you are still believe it insufficient, I can replace the vignette with example codes.  From the view of supporting users, I don't want to do so.

Please give me your feedback.

**Comment by mtmorgan** [comment]
In terms of testing, a modern and comprehensive R package would use formal unit tests to evaluate the logic and implementation of functions and other granular components. R does not really have formal support for 'integration' tests (do the parts of the package work together as advertised?); the vignette is the most important document for this purpose, followed by a lesser extent by examples.

R packages without formal unit tests rely on examples and vignettes for all testing, both 'unit' and 'integration'.

In terms of documentation, examples on man pages are meant to guide the user consulting the man page in how to use the function(s) or data documented on that man page. Vignettes are meant to provide an integrated use of the package over all, including where appropriate how it might be used with other parts of the Bioconductor system.

The vignette therefore serves two essential purposes, as a test of the code over all, and as integrative documentation.

Your static code chunks disable the integration and general test functionality of the vignette. Your vignette almost certainly covers code paths not covered in your examples, in particular when your package is built on different platforms and when there are changes in the code of your package or the packages you depend on. I can only say from observing the Bioconductor build system for the last 12 years that vignettes are invaluable in revealing technical faults in code.

Your static code chunks also change an actual analysis into a kind of fiction, where the pictures of code might have worked at one time or on one computer, but do not work now or on another computer. As a simple case in point, I tried to follow your vignette. Here's the first line I evaluated

```
> download.file("https://superb-sea2.dl.sourceforge.net/project/decd/decd_data.rda",
              "decd_data.rda");
trying URL 'https://superb-sea2.dl.sourceforge.net/project/decd/decd_data.rda'
Error in download.file("https://superb-sea2.dl.sourceforge.net/project/decd/decd_data.rda",  : 
  cannot open URL 'https://superb-sea2.dl.sourceforge.net/project/decd/decd_data.rda'
In addition: Warning message:
In download.file("https://superb-sea2.dl.sourceforge.net/project/decd/decd_data.rda",  :
  cannot open URL 'https://downloads.sourceforge.net/project/decd/decd_data.rda?download&failedmirror=superb-sea2.dl.sourceforge.net': HTTP status was '404 Not Found'
```

I did eventually find https://sourceforge.net/projects/decd/files/?source=navbar where there is a file, but it's called 'input0.rda'. So as a user I can now get nothing with confidence out of your vignette. I downloaded the file anyway, and got to the next code chunk, where I did

```
> hc=hclust(dist(t(exp.brca[, cl==0])))
Error in t(exp.brca[, cl == 0]) : object 'cl' not found
```

Arrgh! Also, this matrix + column metadata format is ideally suited to ExpressionSet or SummarizedExperiment, and I don't understand why it is not in that format.

Replacing your vignette with a collation of examples undermines the documentation role of vignette as an illustration of integrative package use. It also removes the integrative testing component, and likely reduces the amount of code covered by the functions evaluated in the package.

In short, your vignette must have fully evaluated code chunks. The code chunks should take the reader through a real work flow. With some work, I am fully confident that you can identify an appropriate subset of data that is computationally tractable, statistically meaningful, and biologically interesting.

**Comment by menggf** [comment]
@mtmorgan 
I can understand the concerns. Let me find some balance. Update may take one week,

**Comment by menggf** [comment]
@mtmorgan, @nturaga
Please have a look.

**Comment by nturaga** [comment]
Hi @menggf I'll take a look and get back to you

**Comment by nturaga** [comment]
@menggf Please bump the version for another build. 

When i try to run your vignette. Please correct this issue in your vignette. 

```
 > plot(pselfData)
 Error in UseMethod("depth") :
   no applicable method for 'depth' applied to an object of class "NULL"
```

**Comment by nturaga** [comment]
@menggf Your package looks in good shape for acceptance. 

Please fix the easy NOTE and WARNING, flags you have in your build report and bump the version for a fresh build. 

http://bioconductor.org/spb_reports/DEComplexDisease_buildreport_20171107092418.html

**Comment by nturaga** [comment]
@menggf Please let me know if you have fixed the warnings in the build report. I will accept your package then.

**Comment by menggf** [comment]
Hi @nturaga ,
The only warning message is 
"* checking data for ASCII and uncompressed saves ... WARNING
  Warning: package needs dependence on R (>= 2.10)"

During developing our package, our test suggests that this package only works using R > 3.3.3. Otherwise, it may have some problems, e.g. the missed functions.  So I set R > 3.3.3 in DESCRIPTION file. But R check returns a warning for this setting.

Even though this warning message is returned, I don't believe it is critical. Can we just pass it?  Or do you have any suggestion to fix it?

Thanks,

**Comment by nturaga** [comment]
Hi @menggf 

Leaving that issue aside, there are a few more issues with your build report. Please take a look at the `sapply` to `vapply` comment, and also using `seq_len` where needed.

**Comment by menggf** [comment]
hi @nturaga ,
Thank you for reviewing our package and giving suggestions.
I have transformed most of sapply to vapply and used seq_len or seq_along to replace the 1:n.

**Comment by nturaga** [comment]
Hi, 

You still have one left,

https://github.com/menggf/DEComplexDisease/blob/101aa4e420f008d7e9f5262670710c10f82d95cc/R/util.R#L109

**Comment by menggf** [comment]
Hi, @nturaga 
Sorry for missing so many 1:n in previous version. Currently, they are all corrected. 
One "sapply" is left for that  the function will not return a pre-specified type of return value. 'function(x) res.module[[x]][[type]][["patients"]]' returns a flexible-length vector. So, it should leave as sapply.

Best,

**Comment by nturaga** [comment]
Please use the latest version of R ,   `R (>= 3.5.0)` that should take care of the warnings.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/menggf.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `biocLite(\"DEComplexDisease\")`. The package 'landing page' will be created at

https://bioconductor.org/packages/DEComplexDisease

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://docs.google.com/forms/d/e/1FAIpQLSdlTbNjsQJDp0BA480vo4tNufs0ziNyNmexegNZgNieIovbAA/viewform
